src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,148 @@
+                    {
+                        executionContext.Error(stderr.Data);
+                    }
+                };
+
+                var trimChars = new[] { '\\', '/' };
+
+                dropLocation = Path.Combine(dropLocation.TrimEnd(trimChars));
+                downloadFolderPath = downloadFolderPath.TrimEnd(trimChars);
+
+                string robocopyArguments = "\"" + dropLocation + "\" \"" + downloadFolderPath + "\" * /E /COPY:DA /NP /R:3";
+
+                robocopyArguments += " /MT:" + parallelCount;
+
+                int exitCode = await processInvoker.ExecuteAsync(
+                        workingDirectory: "",
+                        fileName: "robocopy",
+                        arguments: robocopyArguments,
+                        environment: null,
+                        requireExitCodeZero: false,
+                        outputEncoding: null,
+                        killProcessOnCancel: true,
+                        cancellationToken: cancellationToken);
+
+                executionContext.Output(StringUtil.Loc("RobocopyBasedPublishArtifactTaskExitCode", exitCode));
+
+                // Exit code returned from robocopy. For more info https://blogs.technet.microsoft.com/deploymentguys/2008/06/16/robocopy-exit-codes/
+                if (exitCode >= 8)
+                {
+                    throw new Exception(StringUtil.Loc("RobocopyBasedPublishArtifactTaskFailed", exitCode));
+                }
+            }
+        }
+
+        private async Task DownloadFileShareArtifactAsync(string sourcePath, string destPath, int parallelCount, CancellationToken cancellationToken, IEnumerable<Func<string, bool>> minimatchFuncs = null)
+        {
+            var trimChars = new[] { '\\', '/' };
+
+            sourcePath = Path.Combine(sourcePath.TrimEnd(trimChars));
+
+            IEnumerable<FileInfo> files =
+                new DirectoryInfo(sourcePath).EnumerateFiles("*", SearchOption.AllDirectories);
+
+            var parallelism = new ExecutionDataflowBlockOptions()
+            {
+                MaxDegreeOfParallelism = parallelCount,
+                BoundedCapacity = 2 * parallelCount,
+                CancellationToken = cancellationToken
+            };
+
+            var actionBlock = NonSwallowingActionBlock.Create<FileInfo>(
+                action: file =>
+                {
+                    if (minimatchFuncs == null || minimatchFuncs.Any(match => match(file.FullName))) 
+                    {
+                        string destinationPath = Path.Combine(destPath, Path.GetRelativePath(sourcePath, file.FullName));
+                        this.context.Output(StringUtil.Loc("CopyFileToDestination", file, destinationPath));
+                        FileInfo destinationFile = new System.IO.FileInfo(destinationPath);
+                        destinationFile.Directory.Create(); // If the directory already exists, this method does nothing.
```

Comment:
Just use Directory.CreateDirectory.

Reply:
Verified this behavior is not what we wanted. It will create the "file" as directory as well!

Reply:
You can grab the destinationPath, call Path.GetDirectoryName and call Directory.CreateDirectory. But this way is fine as well and I kind of prefer this way as it's less lines. It's just semantics.

Reply:
yeah, I know but the current way is also fine. I don't think using Directory.CreateDirectory is more performant.